### PR TITLE
seadrive-gui: init at 1.0.7, seadrive-daemon: init at 1.0.7

### PR DESCRIPTION
**Pull request by justinlovinger**
###### Motivation for this change
SeaDrive is the new client for the Seafile cloud syncing service. Unlike the original `seafile-client`, SeaDrive can download files on-demand, allowing access to your entire Seafile library on devices with low storage space.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

### CODE REVIEW

File: pkgs/applications/networking/seadrive-gui/default.nix
Comment (alexarice) [inline pkgs/applications/networking/seadrive-gui/default.nix]: Do you need a wrapper here instead of just a copy or symlink

**Review by alexarice — COMMENTED** [review]
I don't understand a lot of this and am not sure why you need all the debian libraries but have tried to put some useful comments in

File: pkgs/applications/networking/seadrive-daemon/lib/deblibcurl/default.nix
Comment (symphorien) [inline pkgs/applications/networking/seadrive-daemon/lib/deblibcurl/default.nix]: This kind of url is very short lived. It will disappear in a few weeks. Does nixpkgs's libcurl not work here ?

**Comment by aanderse** [comment]
The website says "open source" and has a github link. Can we not grab the source and compile?

**Comment by JustinLovinger** [comment]
@aanderse seadrive-daemon is not currently open source. seadrive-gui and seafile-client are open source.

File: pkgs/applications/networking/seadrive-daemon/lib/deblibcurl/default.nix
Comment (JustinLovinger) [reply pkgs/applications/networking/seadrive-daemon/lib/deblibcurl/default.nix]: The nixpkgs libcurl is not compatible with the seadrive-daemon binary. Same goes for the other bundled Debian libraries.

File: pkgs/applications/networking/seadrive-gui/default.nix
Comment (JustinLovinger) [reply pkgs/applications/networking/seadrive-gui/default.nix]: `makeWrapper` was recommended over at the #nixos irc. I'm not sure why it was recommended over a symlink.

File: pkgs/applications/networking/seadrive-daemon/lib/libevent_2_0/default.nix
Comment (JustinLovinger) [reply pkgs/applications/networking/seadrive-daemon/lib/libevent_2_0/default.nix]: This libevent_2_0 is actually from a [previous nixpkgs revision](https://github.com/NixOS/nixpkgs/raw/3be19de8e4eba036d1c6ae81295beee1871a7fef~1/pkgs/development/libraries/libevent/default.nix). I haven't touched the code itself.

File: pkgs/applications/networking/seadrive-daemon/default.nix
Comment (JustinLovinger) [reply pkgs/applications/networking/seadrive-daemon/default.nix]: What exactly is set automatically? What would you recommend here?

Comment (aanderse) [reply pkgs/applications/networking/seadrive-daemon/default.nix]: `name = "${pname}-${version}";` is already implied, so as long as you have `pname` and `version` you don't need to declare `name` yourself.

**Review by aanderse — CHANGES_REQUESTED** [review]
Thanks for choosing to contribute to nixpkgs! There are a few things to work through on this PR before we can merge but there are plenty of people who can answer questions you might have.

The initial questions/concerns with this PR for me are all of the custom package versions. Sometimes packages (especially proprietary ones) have fairly specific requirements and custom libs are a part of that, but often times we can work around those limitations using the tools we have at our disposal in nixpkgs.

Please see the questions/comments I have left as a review on this PR and let me know if you require any clarification.

File: pkgs/applications/networking/seadrive-daemon/default.nix
Comment (JustinLovinger) [reply pkgs/applications/networking/seadrive-daemon/default.nix]: Updated. `name` removed.

File: pkgs/applications/networking/seadrive-daemon/lib/libevent_2_0/default.nix
Comment (JustinLovinger) [reply pkgs/applications/networking/seadrive-daemon/lib/libevent_2_0/default.nix]: Updated. `pname` and `version` specified in rec set and `name` removed.

File: pkgs/applications/networking/seadrive-gui/fix_webkit_webengine.patch
Comment (JustinLovinger) [reply pkgs/applications/networking/seadrive-gui/fix_webkit_webengine.patch]: I wrote this patch. I couldn't find any related issues on the seadrive-gui repository, so I assume the issue doesn't exist on other platforms.

File: pkgs/applications/networking/seadrive-gui/default.nix
Comment (JustinLovinger) [reply pkgs/applications/networking/seadrive-gui/default.nix]: [Looks like it defaults to Debug](https://github.com/haiwen/seadrive-gui/blob/master/CMakeLists.txt#L13). I also see that the related `seafile-client` package [sets this flag](https://github.com/NixOS/nixpkgs/blob/f5e4c981fc61be684e6478e209a8b2834b2d6e61/pkgs/applications/networking/seafile-client/default.nix#L22). I will add a comment to clarify.

File: pkgs/applications/networking/seadrive-gui/default.nix
Comment (JustinLovinger) [reply pkgs/applications/networking/seadrive-gui/default.nix]: `seadrive-gui` calls `seadrive-daemon/bin/seadrive` at runtime. I will add a comment.

File: pkgs/applications/networking/seadrive-daemon/lib/libevent_2_0/default.nix
Comment (JustinLovinger) [reply pkgs/applications/networking/seadrive-daemon/lib/libevent_2_0/default.nix]: I don't see any known CVE in libevent 2.0.22. I will add a comment about using libevent 2.0.

File: pkgs/applications/networking/seadrive-daemon/lib/deblibcurl/default.nix
Comment (JustinLovinger) [reply pkgs/applications/networking/seadrive-daemon/lib/deblibcurl/default.nix]: `seadrive-daemon` gives "no version information available" warnings and runtime errors like "libcurl failed to GET ... SSL peer certificate or SSH remote key was not OK." with nixpkgs curl. I have added this explanation in a comment.

File: pkgs/applications/networking/seadrive-daemon/default.nix
Comment (JustinLovinger) [reply pkgs/applications/networking/seadrive-daemon/default.nix]: I have shortened the description.

File: pkgs/applications/networking/seadrive-gui/default.nix
Comment (aanderse) [reply pkgs/applications/networking/seadrive-gui/default.nix]: I haven't checked but I thought [this](https://github.com/NixOS/nixpkgs/blob/7bad92935d12845f31e54f3dc5bafdca3f058ad1/pkgs/development/tools/build-managers/cmake/setup-hook.sh#L93) would take precedence still.

File: pkgs/applications/networking/seadrive-gui/default.nix
Comment (aanderse) [reply pkgs/applications/networking/seadrive-gui/default.nix]: In that case I understand why you were recommended to use `makeWrapper`. Try wrapping the `seadrive-gui` binary and passing the `seadrive-daemon` binary as an argument via `--prefix PATH` like so:
```
wrapProgram $out/bin/seadrive --prefix PATH : ${lib.makeBinPath [ seadrive-daemon ]}
```

File: pkgs/applications/networking/seadrive-gui/default.nix
Comment (JustinLovinger) [reply pkgs/applications/networking/seadrive-gui/default.nix]: Looks like the Release flag is set even without `cmakeFlags`. I have removed that line.

File: pkgs/applications/networking/seadrive-gui/default.nix
Comment (JustinLovinger) [reply pkgs/applications/networking/seadrive-gui/default.nix]: Thank you. Looks like `wrapProgram` works just as well as `makeWrapper`, and doesn't pollute the `seadrive-gui` bin.

**Comment by JustinLovinger** [comment]
@aanderse Do you have any more recommendations? As for the bundled Debian libraries, this was the only way I could find to fix the library mismatch errors. I added the smallest set of bundled libraries to eliminate all errors and warnings.

**Comment by aanderse** [comment]
@JustinLovinger I'm not very good with `patchelf` but wouldn't something like [this](https://github.com/JustinLovinger/nixpkgs/compare/master...aanderse:pr/66540?expand=1) possibly work?

**Comment by JustinLovinger** [comment]
@aanderse It looks like you are trying to trick the program into thinking it has a different version of the library. This was actually discussed as a possible solution on the #nixpkgs irc, but was deemed risky, especially since we are dealing with a data syncing service and don't want to risk losing data.

I just tried running it. It doesn't look like it fixes the warnings

```
./result/bin/seadrive: /nix/store/m93jjbdb9nvc1f38fxbf14rs4r1g013a-curl-7.65.3/lib/libcurl.so.4: no version information available (required by ./result/bin/seadrive)
./result/bin/seadrive: /nix/store/inp0f3gmcq8jxmxqa3spdp9g3zp6rvda-seadrive-daemon-1.0.6/lib/libcrypto.so.1.0.2: no version information available (required by ./result/bin/seadrive)
```

Although the SeaDrive logs don't show any obvious errors, the data itself does not appear to be syncing. No files or Seafile libraries are available.

Also, are you sure replacing the bundled libevent_2_0 with an override like that is a good idea? It seems fragile. The latest libevent could make a change that breaks this libevent_2_0, or worse, breaks seadrive-daemon without breaking libevent_2_0, leading to a difficult to track down problem. With the bundled libevent_2_0 nix, at least we know it will always be the same package, unless we specifically change it.

**Comment by aanderse** [comment]
@JustinLovinger fair enough. I thought I'd at least give it a try.

Who were you working with on IRC? I'd like them to give this a review.

**Comment by JustinLovinger** [comment]
@aanderse The main helpers on irc were srhb, clever, and symphorien. Those were their names on irc.

**Comment by aanderse** [comment]
Ah, great. If @srhb is telling you what to do then it is definitely the right answer. I'll leave the remainder of the review to them.

**Comment by srhb** [comment]
Hi @JustinLovinger! Phew, you sure had to jump through a lot of hoops to make the daemon work. :grin:

I'm sorry to say that I feel that it's a bit *too* many hoops to be currently included in nixpkgs proper, especially considering the potentially high risk libraries like sasl and openssl. I suggest splitting out the open source / built-from-source bits into their own PR and letting the daemon and things live in your own repo such that people may opt into it.

**Comment by JustinLovinger** [comment]
@srhb Given your concerns about these Debian libraries, I took another look at removing them. I tried replacing the Debian binary of `seadrive-daemon` with the CentOS binary. Looks like the CentOS binary has better compatability with nix libraries :D. `seadrive-daemon` can now run using only nix libraries as dependencies.

There is still a warning when `seadrive-daemon` runs

```
/nix/store/7dsisl4gxr16rfa53k8c2grayzdmqz0w-seadrive-daemon-1.0.7/lib/libcrypto.so.10: no version information available (required by /tmp/result/bin/seadrive)
```

but it doesn't seem to result in any runtime errors. I believe the warning is benign, and caused by a lack of, or difference in, an identifier string in `libcrypto.so.1.0.0`

Also, a happy side effect of switching to the CentOS binary is that `seadrive-daemon` is bumped to version 1.0.7. Apparently CentOS has a newer version than Debian.

**Comment by srhb** [comment]
This definitely looks better. However, I'm not sold on the premise that the warning is completely safe. Can you provide some more evidence of this?

Furthermore I'm not sure I like making the old libevent version local to this drv. We might want to keep it in top-level, alongside the newer one. It is still maintained as far as I know.

I'll try and invite some more people to weigh in as well, I don't feel completely comfortable making these calls on my own. :)

**Comment by JustinLovinger** [comment]
@srhb I think I tracked down what is happening with the "no version information available" warning. To begin, this warning is somewhat deceptively named. The warning actually means that an [internal version identifier](https://www.openssl.org/docs/man1.0.2/man3/OPENSSL_VERSION_NUMBER.html), provided through an `SSLeay()` function, does not match what is expected. This warning is [known to occur](https://github.com/openssl/openssl/issues/1631) when you run a binary with an OpenSSL library that it was not built with, because the version identifier includes information such as the [build platform and even when the library was built](https://www.openssl.org/docs/man1.0.2/man3/OPENSSL_VERSION_NUMBER.html). All the warning really tells us is that we are not using the exact copy of OpenSSL that the binary was built with, which is what we want.

Even if the version does match exactly and the built on date and platform is not an issue, we may still get this warning because RHEL (and consequently, CentOS) change the behavior of `SSLeay()` in their OpenSSL libraries, because they believe that ["Some applications perform incorrect (too sensitive) version check and the only way to avoid it is to mask the actual run-time version of OpenSSL built against an old OpenSSL release."](https://bugzilla.redhat.com/show_bug).

Also, I have personally been running this build of SeaDrive without error, and with heavy usage, for the past few days :)

As for `libevent_2_0`, making it a top level package was my first thought as well, so I have no problem with that if it is what people think is best.

**Comment by srhb** [comment]
@JustinLovinger Thanks for tracking that down. That sets my mind at ease a bit. :)

**Comment by JustinLovinger** [comment]
I'm closing this pull request because I'm not using SeaDrive anymore. If anyone wants to pick this up or use SeaDrive themselves, you can find the latest changes here: <https://github.com/JustinLovinger/nixpkgs/tree/init-seadrive>. I give permission for anyone to make a new pull request with these changes, with maintainer changed, and proper attribution of course :) It worked when I last tested it, but that was several months ago.

**Comment by merspieler** [comment]
The closed source seadrive-daemon is no longer needed and has been replaced with the open source seadrive-fuse.
It did build with the following nix config

```
{ fetchFromGitHub
, pkgconfig
, stdenv
, autoreconfHook
, lib
# Package dependencies
, libsearpc
, libselinux
, libuuid
, pcre
, libtool
, libevent
, sqlite
, openssl
, fuse
, vala
, intltool
, jansson
, curl
, python
}:

stdenv.mkDerivation rec {
  pname = "seadrive-fuse";
  version = "2.0.16";

  src = fetchFromGitHub {
    owner = "haiwen";
    repo = pname;
    rev = "v${version}";
    sha256 = "072sx4wvj3gbslv3hn4sifr28fy812b8aja9d7phl1w4yix9l55z";
  };

  patches = [];

  nativeBuildInputs = [
    autoreconfHook
    pkgconfig
  ];
  buildInputs = [
    libsearpc
    libselinux
    libuuid # Satisfies the 'mount' package requirement. Contains 'mount.pc'
    pcre
    libtool
    libevent
    sqlite
    openssl.dev
    fuse
    vala
    intltool
    jansson
    curl
    python
  ];

  meta = with lib; {
    homepage = https://github.com/haiwen/seadrive-fuse;
    description = "SeaDrive daemon with FUSE interface";
    license = licenses.gpl3;
    platforms = platforms.all;
    maintainers = with maintainers; [];
  };
}
```

I couldn't test it tho cause I didn't build the seadrive gui before ending my nixOS test.

**Pull request closed without merging**
